benchmark/single-source/StringReplaceSubrange.swift

```diff
@@ -0,0 +1,116 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import TestsUtils
+
+let tags: [BenchmarkCategory] = [.validation, .api, .String]
+
+public let StringReplaceSubrange = [
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.String",
+    runFunction: { replaceSubrange($0, "coffee", with: "t") },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.String",
+    runFunction: { replaceSubrange($0, largeManagedString, with: "t") },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.Substr",
+    runFunction: { replaceSubrange($0, "coffee", with: getSubstring("t")) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.Substr",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getSubstring("t")) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.ArrChar",
+    runFunction: { replaceSubrange($0, "coffee", with: getArrayCharacter(Array<Character>(["t"]))) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.ArrChar",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getArrayCharacter(Array<Character>(["t"]))) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.RepeatedChar",
+    runFunction: { replaceSubrange($0, "coffee", with: getRepeatedCharacter(repeatedCharacter)) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.RepeatedChar",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getRepeatedCharacter(repeatedCharacter)) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+]
+
+// MARK: - Privates for String
+
+private var largeManagedString: String = {
+    return getString("coffee\u{301}coffeecoffeecoffeecoffee")
+}()
+
+private func setupLargeManagedString() {
+    _ = largeManagedString
+}
+
+// MARK: - Privates for Repeated<Character>
+
+private let repeatedCharacter: Repeated<Character> = {
+  let character = Character("c")
+  return repeatElement(character, count: 1)
+}()
+
+
+@inline(never)
+private func replaceSubrange(_ N: Int, _ string: String, with replacingString: String) {
+  var copy = getString(string)
+  let range = string.startIndex..<string.index(after: string.startIndex)
+  for _ in 0 ..< 5_000 * N {
+    copy.replaceSubrange(range, with: replacingString)
+  }
+}
+
+@inline(never)
+private func replaceSubrange(_ N: Int, _ string: String, with replacingSubstring: Substring) {
```

Comment:
If I understand @milseman's intention from [SR-8905](https://bugs.swift.org/browse/SR-8905) correctly, we are designing benchmark for the generic `replaceSubstring` method that varies across the `String`, `Substring`, `Array<Character>` and `Repeat<Character>` specializations.

Therefore we should be able to define single shared generic test function and vary the parameter in `runFunction` closure in `BenchmarkInfo` declarations.

For an example of such benchmarks, see [`append`](https://github.com/apple/swift/blob/1cbc0b376fdcd6af16d201b35fa568fdf29bb305/benchmark/single-source/DataBenchmarks.swift#L453-L460) variants in [`DataBenchmarks`](https://github.com/apple/swift/blob/1cbc0b376fdcd6af16d201b35fa568fdf29bb305/benchmark/single-source/DataBenchmarks.swift#L233-L238)

@milseman Any thoughts on keeping or dropping the `@inline(never)` annotation?

Reply:
You could refactor this to a generic `@inline(__always)` implementation function that takes a scale factor (because the String version should be much faster than the Repeat version) and does the loop and `replaceSubrange` calls. You'd likely want to make a `@inline(never)` top-level function for each individual benchmark. It doesn't save you a whole lot.